op-chain-ops/cmd/op-version-check/main.go

```diff
@@ -0,0 +1,191 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"slices"
+
+	"golang.org/x/exp/maps"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/mattn/go-isatty"
+	"github.com/urfave/cli/v2"
+
+	"github.com/ethereum-optimism/optimism/op-chain-ops/clients"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/upgrades"
+
+	"github.com/ethereum-optimism/superchain-registry/superchain"
+)
+
+type ChainVersionCheck struct {
+	Name             string                      `yaml:"name"`
+	ChainID          uint64                      `yaml:"chain_id"`
+	ContractVersions superchain.ContractVersions `yaml:"contract_versions"`
+}
+
+func main() {
+	log.Root().SetHandler(log.StreamHandler(os.Stderr, log.TerminalFormat(isatty.IsTerminal(os.Stderr.Fd()))))
+
+	app := &cli.App{
+		Name:  "op-version-check",
+		Usage: "Determine which contract versions are deployed for chains in a superchain",
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:    "l1-rpc-url",
+				Value:   "http://127.0.0.1:8545",
+				Usage:   "L1 RPC URL, the chain ID will be used to determine the superchain",
+				EnvVars: []string{"L1_RPC_URL"},
+			},
+			&cli.Uint64SliceFlag{
+				Name:  "chain-ids",
+				Usage: "L2 Chain IDs corresponding to chains to check versions for. Corresponds to all chains if empty",
+			},
+			&cli.StringFlag{
```

Comment:
Can we replace the `chain-ids` and `superchain-target` flags with a `l2-rpc-urls` flag? This will allow the software to work with L2s that are not part of the superchain. The code to parse the superchain target is brittle and using it in 2 places will create a maintenance burden (its copied from `op-upgrade`).

Could use `StringSliceFlag` - https://github.com/urfave/cli/blob/v2.27.1/flag_string_slice.go

Could also make `l1-rpc-urls` into a string slice flag to make the API symmetric and allow for max flexibility with how the infra team wants to run the software

Using the syntax (# of L1):(# of L2), could run it with 1:1 for a custom network, or 1:many for a superchain or many:0 to observe all L1s

Reply:
Do we really need to check more than one L2 / L1 per invocation?
I think the most complex we'd want is to have a single L1 and multiple L2s

Reply:
Sure but it is minimal to support multiple L1s - if not then the API is not symmetrical which is weird

Reply:
for l2s that are not part of the superchain, how would the addresses and `chainConfig` for checking the versions be obtained? since the chain is not on the superchain, this raises a tricky question @tynes

Reply:
For L2s, the addresses are deterministic, ie can always call the same predeploys

The L1 addresses are not deterministic